api/python/cellxgene_census/src/cellxgene_census/experimental/pp/_highly_variable_genes.py

```diff
@@ -0,0 +1,338 @@
+    )
+    assert pd.api.types.is_categorical_dtype(obs[batch_key].dtype)
+    return obs[batch_key]
+
+
+def _highly_variable_genes_seurat_v3(
+    query: soma.ExperimentAxisQuery,
+    batch_key: Optional[str] = None,
+    n_top_genes: int = 1000,
+    layer: str = "raw",
+    span: float = 0.3,
+) -> pd.DataFrame:
+    try:
+        import skmisc.loess
+    except ImportError as e:
+        raise ImportError("Please install skmisc package via `pip install --user scikit-misc") from e
+
+    batch_indexer = None
+    if batch_key is not None:
+        batch_index = _get_batch_index(query, batch_key)
+        n_batches = len(batch_index.cat.categories)
+        n_samples = batch_index.value_counts().loc[batch_index.cat.categories.to_numpy()].to_numpy()
+        if n_batches > 1:
+            batch_indexer = batch_index.index.get_indexer
+            batch_codes = batch_index.cat.codes.to_numpy().astype(np.int64)
+    else:
+        n_batches = 1
+        n_samples = np.array([query.n_obs], dtype=np.int64)
+
+    assert n_batches == len(n_samples)
+    assert query.n_obs == n_samples.sum()
+    assert all(n_samples > 0)
+    assert (n_batches > 1) == bool(batch_indexer)
+
+    max_workers = (os.cpu_count() or 4) + 2
+    var_indexer = query.indexer
+
+    with futures.ThreadPoolExecutor(max_workers=max_workers) as pool:
+        mvn = MeanVarianceAccumulator(n_batches, n_samples, query.n_vars)
+        for arrow_tbl in EagerIterator(query.X(layer).tables(), pool=pool):
+            data = arrow_tbl["soma_data"].to_numpy()
+            if batch_indexer:
+                _batch_take_at_future = pool.submit(batch_indexer, arrow_tbl["soma_dim_0"])
+                var_dim = var_indexer.by_var(arrow_tbl["soma_dim_1"])
+                _batch_vec = batch_codes[_batch_take_at_future.result()]
+                mvn.update_by_batch(_batch_vec, var_dim, data)
+            else:
+                var_dim = var_indexer.by_var(arrow_tbl["soma_dim_1"])
+                mvn.update_single_batch(var_dim, data)
+
+        batches_u, batches_var, all_u, all_var = mvn.finalize()
+        del mvn
+
+    var_df = pd.DataFrame(
+        index=pd.Index(data=query.var_joinids(), name="soma_joinid"),
+        data={
+            "means": all_u,
+            "variances": all_var,
+        },
+    )
+
+    # Calculate per-batch clip_val and reg_std
+    estimated_variances = np.empty((query.n_vars,), dtype=np.float64)
+    reg_std = np.zeros((n_batches, query.n_vars), dtype=np.float64)
+    clip_val = np.zeros((n_batches, query.n_vars), dtype=np.float64)
+    for batch in range(n_batches):
+        estimated_variances.fill(0)
+        u = batches_u[batch]
+        v = batches_var[batch]
+        N = n_samples[batch]
+
+        not_const = v > 0
+        y = np.log10(v[not_const])
+        x = np.log10(u[not_const])
+        model = skmisc.loess.loess(x, y, span=span, degree=2)
+        # Can we resolve low entropy loess errors by adding jitter and retrying?
```

Comment:
I think this is a good idea. I retry manually with jitter every time I get a loess error, anyway, so it seems reasonable to automate that.

Reply:
OK, will do - it seems simple and very useful.

My feeling is that the parameterization of this should be a new argument called "max_jitter", which can be set to zero to get the old behavior.  The code would retry adding jitter until either success or max_jitter is exceeded.  I'll ping you for another review when it is added.

Reply:
@atarashansky - any thoughts on what would be a reasonable range of jitter values to try before bailing?  It is a 64 bit float, so it can start small.

I'm going to start with  
> [  10 ** -p for p in range(18, 8, -1) ]

i.e.,

> [1e-18, 1e-17, 1e-16, 1e-15, 1e-14, 1e-13, 1e-12, 1e-11, 1e-10, 1e-09]